src/parse/matches/matched_arg.rs

```diff
@@ -124,14 +124,24 @@ impl MatchedArg {
         }
     }
 
-    pub(crate) fn contains_val(&self, val: &str) -> bool {
-        self.vals_flatten().any(|v| {
-            if self.case_insensitive {
-                v.eq_ignore_ascii_case(val)
-            } else {
-                OsString::as_os_str(v) == OsStr::new(val)
-            }
-        })
+    // Provide `None` to `val` if checking for any value
+    // or `Some(...)` if checking for a specific value
+    pub(crate) fn contains_explicit_val(&self, val: Option<&str>) -> bool {
```

Comment:
Making this function fill the roles of
- `contains`
- `contains_val`

By making val an `Option` is a non-straightforward API choice.  This is two separate functions in one.

For example, looking at this call
```rust
matcher.contains_explicit_val(&g.id, None)
```
Reading that straight, it is "the matcher contains  an explicitly set argument that is `None` which doesn't make sense.

Reply:
IIRC we follow this pattern in a few other places in this codebase (including the public API). 

Ex: https://docs.rs/clap/3.0.0-beta.5/clap/struct.Arg.html#method.default_value_if and other 3 related methods.

Reply:
Thanks for the example!

That is a very confusing API.   While its precedence, should we be using that as a model?  Ideally we move away from this as we evolve how we handle these cases, whether its by providing a builder for them or a specialized enum that describes the cases.

Reply:
Actually, that `default_value` `None` looks correct because it is unsetting. Similar to `help_heading` we recently did.

Reply:
`default_value` being `None`, yes.  `val` being `None`, doesn't quite make sense from a straight reading.

As I've been going through my day, I've been going back and forth on how much of an issue this is or isn't.  I think I've come up with a incremental change that will improve the readability here and be a testing bed for public API changes (if we don't do something more radical as part of modularity that can't reuse this).

What if the API is
```rust
pub(crate) ArgPredicate {
    IsPresent,
    Equals(&str),
}

fn check_explicit(&self, predicate: ArgPredicate) -> bool {
    ...
}
```
Not fully set on any of the naming, especially `check_explicit` but it makes the intent clear to the reader and allows for a consistent API between arg and matcher so we can look for one common term.

I have not yet gone through this PR to see how this API would impact it, so no idea if this will work in practice or not :)

Reply:
> be a testing bed for public API changes

I don't mind doing this here because it is definitely an improvement, but I am not sure if we want do it on public API changes (even later on) because it will be another type the user has understand. I would rather us create 2 different methods instead.